storage/SMAIndexSubBlock.hpp

```diff
@@ -0,0 +1,418 @@
+ public:
+  SMAIndexSubBlock(const TupleStorageSubBlock &tuple_store,
+                   const IndexSubBlockDescription &description,
+                   const bool new_block,
+                   void *sub_block_memory,
+                   const std::size_t sub_block_memory_size);
+
+  /**
+   * @brief Frees data associated with variable length attributes.
+   * 
+   * Several of the data structure in this index are in heap, therefore it is
+   * important that the destructor be called when evicted. The variables which
+   * are held out of line are the variable length TypedValues and the comparators. 
+   */
+  ~SMAIndexSubBlock();
+
+  /**
+   * @brief Determine whether an IndexSubBlockDescription is valid for this
+   *        type of IndexSubBlock.
+   *
+   * @param relation The relation an index described by description would
+   *        belong to.
+   * @param description A description of the parameters for this type of
+   *        IndexSubBlock, which will be checked for validity.
+   * @return Whether description is well-formed and valid for this type of
+   *         IndexSubBlock belonging to relation (i.e. whether an IndexSubBlock
+   *         of this type, belonging to relation, can be constructed according
+   *         to description).
+   **/
+  static bool DescriptionIsValid(const CatalogRelationSchema &relation,
+                                 const IndexSubBlockDescription &description);
+
+  /**
+   * @brief Estimate the average number of bytes (including any applicable
+   *        overhead) used to index a single tuple in this type of
+   *        IndexSubBlock. Used by StorageBlockLayout::finalize() to divide
+   *        block memory amongst sub-blocks.
+   * @warning description must be valid. DescriptionIsValid() should be called
+   *          first if necessary.
+   *
+   * @param relation The relation tuples belong to.
+   * @param description A description of the parameters for this type of
+   *        IndexSubBlock.
+   *
+   * @return The average/amortized number of bytes used to index a single
+   *         tuple of relation in an IndexSubBlock of this type described by
+   *         description.
+   **/
+  static std::size_t EstimateBytesPerTuple(const CatalogRelationSchema &relation,
+                                           const IndexSubBlockDescription &description);
+
+  /**
+   * @return The sub block type.
+   */
+  IndexSubBlockType getIndexSubBlockType() const override {
+    return kSMA;
+  }
+
+  /**
+   * @return \c true if the SMA block is initialized.
+   */
+  bool supportsAdHocAdd() const override {
+    return initialized_;
+  }
+
+  /**
+   * @return \c true if the SMA block is initialized.
+   */
+  bool supportsAdHocRemove() const override {
+    return initialized_;
+  }
+
+  /**
+   * @param tuple The new tuple added.
+   * @return \c true always. There's no reason we should ever run out of space once
+   *         this index has been successfully created.
```

Comment:
It is okay, adding a tuple simply marks the SMA as inconsistent. Any queries against the SMA will trigger a rebuild which will make it consistent. It's not a perfect system, but based on the fact that Saket and I will be moving these indexes into a separate block, this detail is likely to change anyways.